[Caffe2] Add Adadelta optimizer to caffe2

sidgoyal78 opened the pull request:
Summary:
- Added CPU/GPU implementations of Adadelta and SparseAdadelta.
- Added corresponding Python unittests

Differential Revision: D8712169

BIT-silence commented on caffe2/sgd/adadelta_op_gpu.cu:
```diff
@@ -0,0 +1,168 @@
+    float gi = g[i];
+    float di = d[i];
+    float hi = nh[i] = decay * h[i] + (1.0f - decay) * gi * gi;
+    float ng = (std::sqrt(di + epsilon) / std::sqrt(hi + epsilon)) * gi;
```
Please do not use std:: functions in kernel. Please use sqrtf and rsqrtf here.
I'd suggest
const float ng =  sqrtf(__ldg(d + i) + epsilon) * rsqrtf(__ldg(nh + i) + epsilon) * __ldg(g + i);

BIT-silence commented on caffe2/sgd/adadelta_op_gpu.cu:
```diff
@@ -0,0 +1,168 @@
+    float* nd,
+    float epsilon,
+    float decay,
+    const float* lr) {
```
Please make the output parameters at the end of the parameter list.
https://google.github.io/styleguide/cppguide.html#Output_Parameters

BIT-silence commented on caffe2/sgd/adadelta_op_gpu.cu:
```diff
@@ -0,0 +1,168 @@
+    float mom_new = mixed_mult(decay, param_mom[paramIdx]) +
+        (1.0f - decay) * grad[gradIdx] * grad[gradIdx];
+    mixed_store(&mom_new, &(param_mom[paramIdx]));
+    float grad_new = (sqrt(mixed_add(epsilon, param_mom_delta[paramIdx])) /
```
same as above. use sqrtf and rsqrtf here.

BIT-silence commented on caffe2/sgd/adadelta_op.h:
```diff
@@ -0,0 +1,179 @@
+
+namespace caffe2 {
+
+template <typename Context>
```
Make such internal function in an unnamed namespace. And also please make the function names follow https://google.github.io/styleguide/cppguide.html#Function_Declarations_and_Definitions

sidgoyal78 replied:
Sorry i didn't get this. I think I followed the template as in other ops (like adagrad).

BIT-silence commented on caffe2/sgd/adadelta_op.h:
```diff
@@ -0,0 +1,179 @@
+  }
+
+ protected:
+  T epsilon_;
```
const float epsilon_

BIT-silence commented on caffe2/sgd/adadelta_op.h:
```diff
@@ -0,0 +1,179 @@
+
+ protected:
+  T epsilon_;
+  T decay_;
```
const float decay_;

BIT-silence commented on caffe2/sgd/adadelta_op.h:
```diff
@@ -0,0 +1,179 @@
+  USE_OPERATOR_CONTEXT_FUNCTIONS;
+  AdadeltaOp(const OperatorDef& operator_def, Workspace* ws)
+      : Operator<Context>(operator_def, ws),
+        epsilon_(OperatorBase::GetSingleArgument<T>("epsilon", 1e-5f)),
```
consider to use OP_SINGLE_ARG

BIT-silence commented on caffe2/sgd/adadelta_op.h:
```diff
@@ -0,0 +1,179 @@
+        float di = momentDeltaIn[idx];
+        float hi = momentOut[idx] =
+            decay_ * momentIn[idx] + (static_cast<T>(1) - decay_) * gi * gi;
+        float ng = (std::sqrt(di + epsilon_) / std::sqrt(hi + epsilon_)) * gi;
```
use std::rsqrt here . std::sqrt is slower than it and actually there is no need to use such divide here.

BIT-silence commented on caffe2/sgd/adadelta_op.h:
```diff
@@ -0,0 +1,179 @@
+    float gi = g[i];
+    float di = d[i];
+    float hi = nh[i] = decay * h[i] + (1.0f - decay) * gi * gi;
+    float ng = (std::sqrt(di + epsilon) / std::sqrt(hi + epsilon)) * gi;
```
use std::rsqrt here

BIT-silence commented on caffe2/sgd/adadelta_op.h:
```diff
@@ -0,0 +1,179 @@
+  for (auto i = 0; i < N; ++i) {
+    float gi = g[i];
+    float di = d[i];
+    float hi = nh[i] = decay * h[i] + (1.0f - decay) * gi * gi;
```
Split it into two separate lines.

BIT-silence commented on caffe2/sgd/adadelta_op.h:
```diff
@@ -0,0 +1,179 @@
+    float decay,
+    const float* lr,
+    Context* /*context*/) {
+  for (auto i = 0; i < N; ++i) {
```
I'd suggest to write this function like.

ConstEigenVectorArrayMap<T> w_arr(w, N);
ConstEigenVectorArrayMap<T> g_arr(g, N);
ConstEigenVectorArrayMap<T> h_arr(h, N);
ConstEigenVectorArrayMap<T> d_arr(d, N);
EigenVectorArrayMap<T> ng_arr(nw, N);
EigenVectorMap<T>(nh, N) = decay * h_arr + (T(1) - decay) * g_arr.square();
ng_arr = (d_arr + epsilon).sqrt() * (h_arr + epsilon).rsqrt() * g_arr;
EigenVectorMap<T>(nd, N) = decay * d_arr + (T(1) - decay) * ng_arr.square();
EigenVectorMap<T>(nw, N) = w_arr + lr[0] * ng_arr;

It will be much faster on CPU to use eigen here since most of the ops above in Eigen have SIMD support.
https://eigen.tuxfamily.org/dox/group__CoeffwiseMathFunctions.html

BIT-silence commented on caffe2/sgd/adadelta_op.h:
```diff
@@ -0,0 +1,179 @@
+#pragma once
```
Please use #ifndef instead of #pragma once to guard the h file.
https://google.github.io/styleguide/cppguide.html#The__define_Guard

BIT-silence commented on caffe2/sgd/adadelta_op_gpu.cu:
```diff
@@ -0,0 +1,168 @@
+        this, Input(PARAM));
+  }
+
+  template <typename IndexType, typename THalf>
```
I'm a little confused about the template here. Since on the Op-level we have a template type T, how is the relationship between T and THalf. I think it's better to remove T at Op-level if you want to use DoRunWithType.

sidgoyal78 replied:
For your comments regarding T and THalf, since we don't have kernels for other types like float16 or double, I have removed the typename T at Op-level, for simplicity. However, even when we had type T at the Op level, I don't think there would be an issue.

Let me make sure that what I understand is reasonable (so that we are on the same page). I think this is what is happening with the current implmentation:
- Since DispatchHelper in L109 (`return DispatchHelper<TensorTypes<int32_t, int64_t>>::call(this, Input(INDICES))`, would produce something the nested if-else code like the following:
```
if (this->INPUT(INDICES).template IsType<int32_t>()) {
        DoRunWithType<int32_t>();
    } else if (this->INPUT(INDICES).template IsType<int32_t>()) {
        DoRunWithType<int64_t>();
    }
```
Now we have `DispatchHelper<TensorTypes2<float, float16>` in the `DoRunWithType()`. This again will call `DoRunWithType2()` with `<float>` or `<float16>` depending on the type of `Input(PARAM)`.

So even when we had type T at the Op level, only `const auto* gradIn = Input(GRAD).template data<T>();` is different (L127). So this would have been still fine.

BIT-silence replied:
Currently it looks much better, thanks.

BIT-silence commented on caffe2/sgd/adadelta_op_gpu.cu:
```diff
@@ -0,0 +1,168 @@
+            grad_slice_sz,
+            epsilon_,
+            decay_,
+            Output(OUTPUT_PARAM)->template mutable_data<THalf>(),
```
paramOut

sidgoyal78 replied:
Fixed.

BIT-silence commented on caffe2/sgd/adadelta_op_gpu.cu:
```diff
@@ -0,0 +1,168 @@
+            epsilon_,
+            decay_,
+            Output(OUTPUT_PARAM)->template mutable_data<THalf>(),
+            Output(OUTPUT_MOMENT_GRAD)->template mutable_data<THalf>(),
```
mementOut

BIT-silence commented on caffe2/sgd/adadelta_op_gpu.cu:
```diff
@@ -0,0 +1,168 @@
+            decay_,
+            Output(OUTPUT_PARAM)->template mutable_data<THalf>(),
+            Output(OUTPUT_MOMENT_GRAD)->template mutable_data<THalf>(),
+            Output(OUTPUT_MOMENT_DELTA)->template mutable_data<THalf>(),
```
momentDeltaOut

BIT-silence commented on caffe2/sgd/adadelta_op.h:
```diff
@@ -0,0 +1,179 @@
+    }
+
+    auto block_size = Input(GRAD).size() / n;
+    for (auto i = 0; i < n; ++i) {
```
Please do not use auto for simple types. int i = 0 or std::size_t i = 0 is fine. auto will make things confused here.

BIT-silence commented on caffe2/sgd/adadelta_op_gpu.cu:
```diff
@@ -0,0 +1,168 @@
+#include "adadelta_op.h"
```
#include "caffe2/sgd/adadelta_op.h"

https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes

BIT-silence commented on caffe2/sgd/adadelta_op_gpu.cu:
```diff
@@ -0,0 +1,168 @@
+    float decay,
+    const float* lr) {
+  CUDA_1D_KERNEL_LOOP(i, N) {
+    float gi = g[i];
```
there is no need to capture g[i] here. Prefer to use __ldg

BIT-silence commented on caffe2/sgd/adadelta_op_gpu.cu:
```diff
@@ -0,0 +1,168 @@
+  USE_OPERATOR_CONTEXT_FUNCTIONS;
+  CUDASparseAdadeltaOp(const OperatorDef& operator_def, Workspace* ws)
+      : Operator<Context>(operator_def, ws),
+        epsilon_(OperatorBase::GetSingleArgument<T>("epsilon", 1e-5f)),
```
Here we can make epsilon and decay as float type and cast it to T when using it.

sidgoyal78 reviewed (commented):
Thanks for reviewing. I have modified the implementation accordingly.

BIT-silence commented:
@pytorchbot retest this please

xiaomengy reviewed (approved):
LGTM. Thanks for this effort. I will take care of the next steps for eigen and CUDA optimization

sidgoyal78 commented:
@pytorchbot retest this please

sidgoyal78 commented:
@pytorchbot retest this please

BIT-silence commented:
@pytorchbot retest this please

sidgoyal78 committed:
Add Adadelta optimizer to caffe2 (#9088)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/9088

Closes https://github.com/pytorch/pytorch/pull/9088

- Added CPU/GPU implementations of Adadelta and SparseAdadelta.
- Added corresponding Python unittests

Differential Revision: D8712169

fbshipit-source-id: f42f79df0038a100f17b900ac676d080b2fce9a0
